src/core/net/ip6.cpp

```diff
@@ -948,7 +948,7 @@ Error Ip6::HandlePayload(Message &          aMessage,
     Error    error   = kErrorNone;
     Message *message = nullptr;
 
-    VerifyOrExit(aIpProto == kProtoUdp || aIpProto == kProtoIcmp6);
+    OT_ASSERT(aIpProto != kProtoTcp || aMessageOwnership != Message::kCopyToUse);
```

Comment:
```suggestion
    VerifyOrExit(aIpProto == kProtoUdp || aIpProto == kProtoIcmp6, {
      if (aMessageOwnership == Message::kTakeCustody) {
          message = &aMessage;
      }
    });
```

I prefer slightly not to use `OT_ASSERT` here. I think this method is tend to be general to handle all kinds of situations. 

Thoughts? @abtink

Reply:
@gjc13 , very good catch!

I fully agree with @simonlingoogle . `OT_ASSERT()` is not good here and the code here should be general.

I think would be even simpler to set the `message` at the start of this method instead of in `VerifyOrExit()` 
- In general, I feel using braces `{ }` in `VerifyOrExit()` make it less readable, and better to use `if() { ....;  ExitNow(); }` for this. 

Here is my suggestion for the change:
```c++
Error Ip6::HandlePayload(Message &          aMessage,
                         MessageInfo &      aMessageInfo,
                         uint8_t            aIpProto,
                         Message::Ownership aMessageOwnership)
{
    Error    error   = kErrorNone;
    Message *message = (aMessageOwnership == Message::kTakeCustody) ? &aMessage : nullptr;

    VerifyOrExit(aIpProto == kProtoUdp || aIpProto == kProtoIcmp6);

    if (message == nullptr) // `aMessageOwnership` is `kCopyToUse`
    {
        VerifyOrExit((message = aMessage.Clone()) != nullptr, error = kErrorNoBufs);
    }
```

Basically we set the `message` at the top of the method based on the `aMessageOwnership` and then do the `VerifyOrExit()` check. If we see `message` is `nullptr` it indicates we need to copy.

Reply:
I guess @gjc13 just wants to assert that TCP will not be handled and forwarded at the same time. Because TCP doesn't work on multicast. If we assert here, attackers may turn a node down by sending TCP traffic with a multicast destination.

Reply:
Fully agree with @abtink. Actually, had almost the same code in mind.

Reply:
Good catch! Thank you, @gjc13 !

I agree with @simonlingoogle and @abtink . Though currently calls with `aIpProto == kProtoTcp && aMessageOwnership == kCopyToUse` only happen with multicast TCP messages (which shouldn't exist), there's no guarantee that the node would never receive such messages.

Reply:
Updated.